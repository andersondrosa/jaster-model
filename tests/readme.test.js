import Model from "../src";
import { object, rm } from "../src";
import { order, nested, deepNested, hash } from "../src/utils";
import firedata from "../lib/firedata";

// Log data as pretty JSON
const log = (text, o) => console.log(text, JSON.stringify(o, null, 2));

it("README - Link and Zelda", function() {
  //
  const character = {
    name: "Character",
    type: "villager",
    rupees: 30,
    gender: "male"
  };

  const link = new Model(character);

  link.put({
    name: "Link"
  });

  // So to see the link name just run the get function
  // log("link name", link.get("name"));
  expect(link.get("name")).toMatchSnapshot();
  // Link, no longer Character

  // But the character type is still Villager
  // log("link type", link.get("type")); // villager
  expect(link.get("type")).toMatchSnapshot();

  // Then, let's change the type of the link to warrior 💪🏻
  link.put({ type: "warrior 🧝🏼‍♂" });

  // And also increase his rupees into 100 😏💎
  link.put(state => {
    return { rupees: state.rupees + 100 };
  });

  // let's see how our character became?
  // log("data", link.getData());
  expect(link.getData()).toMatchSnapshot();

  // Now we can save the changes by making the commits
  link.commit();

  expect(link.getCommits()).toMatchSnapshot();

  // Now that it has been saved, if something happens to the link
  // just restore the data for the last commit

  // Oops, some zombie infected the link and changed its name 😱
  link.put(({ name }) => ({
    name: `Evil '${name}' 🧟‍♂`
  }));

  // log("link name", link.get("name")); // Evil 'Link' 🧟‍♂
  expect(link.get("name")).toMatchSnapshot();

  // Lets try to restoring removing the latest changes
  link.discardChanges();

  // Phew.. the link went back to normal 😆🙌🏻
  // log("link name", link.get("name")); // Link
  expect(link.get("name")).toMatchSnapshot();

  // So, I think we better save it somewhere so we don't lose the link again
  let { data, commits } = link.exportData();

  // And we can take a look to see if everything is alright
  // log("Link initial data", data);
  expect(data).toMatchSnapshot();

  // And see our first commit too
  // log("Link commits", commits);
  expect(commits).toMatchSnapshot();

  //The exported data can reconstruct the link to edit later if we need to,
  // however, if you don't need to edit it anymore,
  // just get your data in a simple way
  expect(link.getData()).toMatchSnapshot();

  // ===========================================================================

  // What do you think about using the link as a model to create the Zelda princess?

  // So we need to create a simple model based on the saved data
  const zelda = new Model(data, commits);

  // We must then set the princess's values and (of course)
  // multiply her rupees by 20 times
  zelda.put(state => {
    return {
      name: "Zelda",
      type: "princess 🧝🏼‍♀",
      gender: "female",
      rupees: state.rupees * 20,
      // And we can use her crush data here 😏
      wholoves: {
        name: state.name + " ♥",
        type: state.type + " 💪🏻"
      }
    };
  });

  // Then we make commit again
  zelda.commit();

  // console.log(zelda.getCommits());

  // Now just start your adventure game
  // log("data", zelda.getData());
  expect(zelda.getData()).toMatchSnapshot();

  expect(link.exportData()).toMatchSnapshot();

  expect(zelda.exportData()).toMatchSnapshot();

  // ===========================================================================

  zelda.use(firedata);

  zelda.put({
    $set: {
      stage: "zelda"
    },
    "@match:next": {
      xp: "required|gte:12"
    }
  });

  zelda.commit();

  zelda.$submit();

  zelda.put({ xp: 12 });

  zelda.$submit();

  const onChange = (objToWatch, event) => {
    const handler = {
      get(target, property, receiver) {
        event();
        const value = Reflect.get(target, property, receiver);
        if (typeof value === "object") {
          return new Proxy(value, handler);
        }
        return value;
      },
      set(target, property, value) {
        event();
        return Reflect.set(target, property, value);
      },
      deleteProperty(target, property) {
        event();
        return Reflect.deleteProperty(target, property);
      }
    };
    return new Proxy(objToWatch, handler);
  };

  class ninja {
    constructor() {
      this.teste = {
        ninja: "oi"
      };
    }
  }

  const $a = onChange(new ninja(), function name(params) {
    console.log("CHANGE");
  });

  console.log($a instanceof ninja);
});
