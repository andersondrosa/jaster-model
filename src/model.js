import { get, isObject, order, clone, nested, emptyObject, hash } from "./utils.js";
import { merge, diff, sha256 } from "jaster";

function setStartData($this, data) {
  $this.startData = $this.commitData = JSON.stringify(order(data));
  $this.data = JSON.parse($this.startData);
}

function setCommitData($this) {
  $this.commitData = isObject($this.data)
    ? JSON.stringify(order($this.data))
    : "{}";
  $this.dirty = false;
}

export default class {
  //
  constructor(data, commits) {
    //
    this.data = {};
    this.startData = "{}";
    this.commitData = "{}";
    this.undoCommits = [];
    this.commits = [];
    this.dirty = false;

    this.prevData = [];
    this.nextData = [];

    if (data && isObject(data)) {
      setStartData(this, data);
    }

    if (Array.isArray(commits)) {
      this.setCommits(commits, true);
    }
  }

  /**
   * Setando os dados iniciais que serão usados de base caso precise restaurar
   * ou desfazer algum comit, por é necessário usar a base e re mesclar todos
   * os comits sobre a base
   */

  // ===========================================================================

  // GETTERS

  hasCommits() {
    return this.commits.length > 0;
  }

  get(key) {
    return get(key, this.data);
  }

  getCommits() {
    return this.commits.map(row => clone(row));
  }

  getCommitData() {
    if (!this.commitData) {
      return {};
    }
    return JSON.parse(this.commitData);
  }

  discardChanges() {
    //
    this.data = this.getCommitData();

    this.cleanHistoric();
  }

  getData() {
    if (this.data) {
      return order(this.data);
    }
    return {};
  }

  exportData() {
    return {
      data: JSON.parse(this.startData),
      commits: this.getCommits()
    };
  }

  getIntegrity() {
    return hash(this.commitData);
  }

  getUncommitedData() {
    //
    if (this.dirty == false) {
      return {};
    }

    return order(diff(this.getCommitData(), this.data));
  }

  commitRequired() {
    //
    if (this.dirty == false) {
      return false;
    }

    if (this.commitData == JSON.stringify(this.getData())) {
      this.dirty = false;
    }

    return this.dirty;
  }

  // ===========================================================================
  // ===========================================================================

  // ACTIONS

  setCommits(commits) {
    this.commits = clone(commits);
    this.restore();
  }

  cleanHistoric() {
    this.prevData = [];
    this.nextData = [];
  }

  put(data) {
    //
    if (data instanceof Function) {
      data = data(this.data);
    }

    const info = merge(this.data, data);

    if (!info.unchanged) {
      this.prevData.push(info);
    }

    this.nextData = [];

    this.dirty = true;

    return this;
  }

  rollback(n = 1) {
    //
    for (let i = 0; i < n; i++) {
      const row = this.prevData.pop();

      if (row == undefined) {
        return false;
      }

      this.nextData.push(merge(this.data, row.rollback));

      this.dirty = true;
    }

    return !!this.dirty;
  }

  rollforth(n = 1) {
    //
    for (let i = 0; i < n; i++) {
      const row = this.nextData.pop();

      if (row == undefined) {
        return false;
      }

      this.prevData.push(merge(this.data, row.rollback));

      this.dirty = true;
    }

    return !!this.dirty;
  }

  // ===========================================================================
  // ===========================================================================

  validateIntegrity(fdata) {
    //
    let certified;

    if (fdata.hasOwnProperty("_certified")) {
      certified = fdata._certified;
      fdata._certified = "<hidden>";
    }

    const str_data = JSON.stringify(fdata);

    const integrity = hash(str_data);

    if (certified) {
      fdata._certified = certified;

      if ((certified === integrity) == false) {
        throw "Invalid certified " +
          certified +
          ". Valid certified must match " +
          integrity;
      }
    }
  }

  commit(message, type = "put") {
    //
    if (this.dirty == false) {
      return false;
    }

    const data = this.getUncommitedData();

    if (emptyObject(data)) {
      this.dirty = false;
      return false;
    }

    const integrity = this.getIntegrity();

    const commit = {
      type,
      message,
      integrity,
      data
    };

    if (!message) {
      delete commit.message;
    }

    this.commits.push(commit);

    const rollback = diff(this.data, this.getCommitData());

    this.undoCommits.push(rollback);

    setCommitData(this);

    return true;
  }

  cleanCommits() {
    //
    this.commits = [];

    this.restore();
  }

  rollbackCommit(levels = 1) {
    //
    this.data = this.getCommitData();

    for (let i = 0; i < levels; i++) {
      this.commits.pop();
      const undo = this.undoCommits.pop();
      merge(this.data, undo);
    }

    setCommitData(this);
  }

  restore() {
    //
    this.data = JSON.parse(this.startData);

    this.undoCommits = this.commits.map(({ data }) => {
      const { rollback } = merge(this.data, data);
      return rollback;
    });

    setCommitData(this);
  }

  // ===========================================================================
  // ===========================================================================
  // ===========================================================================

  // ALIAS

  putn(o) {
    return this.put(nested(o, true));
  }
}
