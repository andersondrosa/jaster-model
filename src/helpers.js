"use strict";

export function isObject(o) {
  return o instanceof Object && o.constructor.name == "Object";
}

export function emptyObject(o) {
  for (let i in o) {
    return false;
  }
  return true;
}

export function get(path, data) {
  let keys = path.split(".");

  let pointer = data;

  let _key = keys.pop();

  for (let i in keys) {
    let key = keys[i];

    if (typeof data != "object" || !pointer.hasOwnProperty(key)) {
      return;
    }

    pointer = pointer[key];
  }

  return pointer[_key] ? pointer[_key] : null;
}

export function lget(path, layers) {
  if (typeof layers != "object") {
    return;
  }

  if (path.indexOf("[") > -1) {
    path = path.replace(/\[([^\]]+)\]/g, (m, i) => {
      return "." + i;
    });
  }

  let i = path.indexOf(".");

  let root = i == -1 ? path : path.slice(0, i);

  let data;

  for (let i = layers.length - 1; i >= 0; i--) {
    if (layers[i].hasOwnProperty(root)) {
      data = layers[i][root];
      break;
    }
  }

  if (i == -1) {
    return data;
  }

  if (typeof data == "object") {
    return this.get(path.slice(i + 1), data);
  }
}

export function slget(key, layers, def) {
  if (typeof layers != "object") {
    return null;
  }

  for (let i = layers.length - 1; i >= 0; i--) {
    if (layers[i].hasOwnProperty(key)) {
      return layers[i][key];
    }
  }

  return def;
}

export function set(data, path, value) {
  let keys = path.split(".");

  let pointer = data;

  let _key = keys.pop();

  for (let i in keys) {
    let key = keys[i];

    if (!pointer.hasOwnProperty(key) || !isObject(pointer[key])) {
      pointer[key] = {};
    }

    pointer = pointer[key];
  }

  pointer[_key] = value;
}

export function parsePath(path, key) {
  if (path) {
    path = path.split(".");
  } else {
    path = [];
  }

  let keys = key.split("/");

  let realkey = keys.pop();

  for (let i in keys) {
    if (keys[i] == "root") {
      path = [];
      break;
    }

    if (keys[i] == "..") {
      path.pop();
    }
  }

  path.push(realkey);

  let realpath = path.join(".");

  return realpath;
}

export function size(object) {
  var length = 0;
  for (var key in object) {
    if (object.hasOwnProperty(key)) {
      ++length;
    }
  }
  return length;
}
