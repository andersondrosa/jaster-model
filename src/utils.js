"use strict";

export function isObject(o) {
  return o instanceof Object && o.constructor.name == "Object";
}

export function emptyObject(o) {
  for (let i in o) {
    return false;
  }
  return true;
}

export function get(path, data) {
  let keys = path.split(".");

  let pointer = data;

  let _key = keys.pop();

  for (let i in keys) {
    let key = keys[i];

    if (typeof data != "object" || !pointer.hasOwnProperty(key)) {
      return;
    }

    pointer = pointer[key];
  }

  return pointer[_key] ? pointer[_key] : null;
}

export function lget(path, layers) {
  if (typeof layers != "object") {
    return;
  }

  if (path.indexOf("[") > -1) {
    path = path.replace(/\[([^\]]+)\]/g, (m, i) => "." + i);
  }

  let i = path.indexOf(".");

  let root = i == -1 ? path : path.slice(0, i);

  let data;

  for (let i = layers.length - 1; i >= 0; i--) {
    if (layers[i].hasOwnProperty(root)) {
      data = layers[i][root];
      break;
    }
  }

  if (i == -1) {
    return data;
  }

  if (typeof data == "object") {
    return this.get(path.slice(i + 1), data);
  }
}

export function slget(key, layers, def) {
  if (typeof layers != "object") {
    return null;
  }

  for (let i = layers.length - 1; i >= 0; i--) {
    if (layers[i].hasOwnProperty(key)) {
      return layers[i][key];
    }
  }

  return def;
}

export function set(data, path, value) {
  const keys = path.split(".");
  let key;
  let pointer = data;
  let _key = keys.pop();
  for (let i in keys) {
    key = keys[i];
    if (!pointer.hasOwnProperty(key) || !isObject(pointer[key])) {
      pointer[key] = {};
    }
    pointer = pointer[key];
  }
  pointer[_key] = value;
}

export function parsePath(path, key) {
  if (path) {
    path = path.split(".");
  } else {
    path = [];
  }

  let keys = key.split("/");

  let realkey = keys.pop();

  for (let i in keys) {
    if (keys[i] == "root") {
      path = [];
      break;
    }

    if (keys[i] == "..") {
      path.pop();
    }
  }

  path.push(realkey);

  let realpath = path.join(".");

  return realpath;
}

export function size(object) {
  var length = 0;
  for (var key in object) {
    if (object.hasOwnProperty(key)) {
      ++length;
    }
  }
  return length;
}

function order(o) {
  const r = Array.isArray(o) ? [] : {};
  let v, k, i;
  const keys = Object.keys(o).sort();
  for (i in keys) {
    k = keys[i];
    v = o[k];
    r[k] = typeof v === "object" && v !== null ? order(v) : v;
  }
  return r;
}

function clone(o) {
  let r, v, k;
  if (typeof o !== "object" || o === null) {
    return o;
  }
  r = Array.isArray(o) ? [] : {};
  for (k in o) {
    v = o[k];
    r[k] = typeof v === "object" && v !== null ? clone(v) : v;
  }
  return r;
}

function nested(o, deep) {
  if (deep) {
    return deepNested(o);
  }
  const r = {};
  for (const k in o) {
    set(r, k, o[k]);
  }
  return r;
}

function deepNested(o) {
  //
  if (Array.isArray(o)) {
    return o.map(v => (v && typeof v == "object" ? deepNested(v) : v));
  }

  let v;
  const r = {};
  for (const k in o) {
    v = o[k];
    set(r, k, v && typeof v == "object" ? deepNested(v) : v);
  }
  return r;
}

export { order, clone, nested };
