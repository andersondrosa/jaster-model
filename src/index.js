import { get, isObject } from "./helpers.js";
import { merge, diff, sha256 } from "jaster";

export default class {
  //

  constructor(data, commits) {
    //
    this.data;
    this.baseData;
    this.commitData;

    this.state = {
      commits: []
    };

    this.prevData = [];
    this.nextData = [];

    this.undoCommits = [];

    if (isObject(data)) {
      this.setBaseData(data);
    }

    if (Array.isArray(commits)) {
      this.setCommits(commits, true);
    }
  }

  /**
   * Setando os dados iniciais que serão usados de base caso precise restaurar
   * ou desfazer algum comit, por é necessário usar a base e re mesclar todos
   * os comits sobre a base
   */
  setBaseData(data) {
    this.baseData = JSON.stringify(data);
    this.data = JSON.parse(this.baseData);
    // SE commitData for NULL, é para se usar o baseData
    this.commitData = null;
  }

  commitRequired() {
    return this.prevData.length > 0;
  }

  cleanHistoric(data) {
    this.prevData = [];
    this.nextData = [];
  }

  isPristine() {
    return !(this.state.commits.length > 0);
    // return JSON.stringify(this.data) == "{}" && this.prevData.length == 0;
  }

  put(data) {
    //
    if (this.data == null) {
      this.data = data;
      this.registerData();
      return this;
    }

    if (data instanceof Function) {
      data = data(this.data);
    }

    let info = merge(this.data, data);

    if (!info.unchanged) {
      this.prevData.push(info);
    }

    this.nextData = [];

    return this;
  }

  rollback(n = 1) {
    var l = this.prevData.length;

    for (let i = 0; i < n; i++) {
      const row = this.prevData.pop();

      if (row == undefined) {
        return;
      }

      const info = merge(this.data, row.rollback);

      this.nextData.push(info);
    }
  }

  rollforth(n = 1) {
    var l = this.nextData.length;

    for (let i = 0; i < n; i++) {
      const row = this.nextData.pop();

      if (row == undefined) {
        return;
      }

      const info = merge(this.data, row.rollback);

      this.prevData.push(info);
    }
  }

  get(path) {
    return get(path, this.data);
  }

  getPath(data, path) {
    //
    let paths = [];

    if (path) {
      for (const key in data) {
        if (data[key].constructor === Object) {
          paths = paths.concat(this.getPath(data[key], path + "." + key));
          continue;
        }

        paths.push(path + "." + key);
      }

      return paths;
    }

    for (const key in data) {
      if (data[key].constructor === Object) {
        paths = paths.concat(this.getPath(data[key], key));
        continue;
      }

      paths.push(key);
    }

    return paths;
  }

  // ===========================================================================
  // ===========================================================================

  validateIntegrity(fdata) {
    let certified;

    if (fdata.hasOwnProperty("_certified")) {
      certified = fdata._certified;
      fdata._certified = "<hidden>";
    }

    let str_data = JSON.stringify(fdata);

    let hash = sha256(str_data);

    if (certified) {
      fdata._certified = certified;

      if ((certified === hash) == false) {
        throw "Invalid certified " +
          certified +
          ". Valid certified must match " +
          hash;
      }
    }
  }

  getCommits() {
    return this.state.commits;
  }

  commit(message) {
    //
    let commitData = diff(this.getComparisonData(), this.data);

    this.addCommit("PUT", message, commitData);

    return this;
  }

  addCommit(type, message, data) {
    //
    const commit = { type, message, data };

    message == undefined && delete commit.message;

    this.state.commits.push(commit);

    this.cleanHistoric();

    this.registerData();
  }

  cleanCommits() {
    //
    this.state.commits = [];
  }

  async rollbackCommit(levels, cb) {
    //
    let state = this.state;
    let commits = [];

    levels = state.commits.length - levels;

    for (const i in state.commits) {
      if (i == levels) {
        state.commits = commits;
        return;
      }

      const commit = state.commits[i];

      commits.push(commit);
    }

    this.cleanHistoric();

    state.commits = commits;

    await this.restore();

    cb && cb();
  }

  registerData() {
    this.commitData = JSON.stringify(this.data);
  }

  getComparisonData() {
    if (!this._data) {
      this._data = JSON.parse(this.commitData);
    }
    return this._data;
  }

  getData(utils) {
    return JSON.parse(JSON.stringify(this.data));
  }

  async getExtensionData(ext) {
    //
    if (!this.$fdata.dependencyProvider) {
      throw "Dependency Provider must be defined";
    }

    let data = {};

    let fdata = await this.$fdata.loadResource(ext);

    if (fdata._extends) {
      data = await this.getExtensionData(fdata._extends);
    }

    if (fdata._commits) {
      for (let i in fdata._commits) {
        merge(data, fdata._commits[i].data);
      }
    }

    return data;
  }

  async restore() {
    //
    const state = this.state;

    this.data = JSON.parse(this.baseData);

    for (const i in state.commits) {
      //
      const commit = state.commits[i];

      this.undoCommits.push(merge(this.data, commit.data));
    }

    this.registerData();
  }

  commitRequired() {
    //
    let request = { response: false };

    this.emmit("commit-required", [request]);

    return request.response == true;
  }

  setCommits(commits, restore) {
    this.state.commits = commits;
    restore == true && this.restore();
  }
}
