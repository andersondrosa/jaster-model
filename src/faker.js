import { nested, order } from "./utils";
const lorem = [
  "alpha",
  "beta",
  "gamma",
  "delta",
  "zeta",
  "lambda",
  "pi",
  "sigma",
  "omega",
  "sampi"
]

function getRandom(arr) {
  return arr[rand(arr.length - 1)];
}
function rand(num) {
  return Math.floor(Math.random() * (num + 1));
}

function fakerValue() {
  return getRandom(lorem);
}

function randLine() {
  let line = [];
  const l = rand(4);
  for (let i = 0; i < l; i++) {
    line.push(getRandom(lorem));
  }
  if(line.length == 0) {
    return getRandom(lorem)
  }
  return line.join(".");
}

function reduce(res, values) {
  const num = values.shift();
  if (!(num > 0)) {
    return res;
  }
  let response = {};
  let i = 0;
  for (let key in res) {
    if (i >= num) break;
    i++;
    if (typeof res[key] == "object" && values.length > 0) {
      response[key] = reduce(res[key], values.slice());
    } else {
      response[key] = res[key];
    }
  }
  return response;
}

function faker() {
  //
  let res = {};

  for (let i = 0; i < 100000; i++) {
    res[randLine()] = fakerValue();
    getRandom(lorem);
  }

  res = nested(res);

  const response = {
    alpha: reduce(res.alpha, [3, 2, 5]),
    beta: reduce(res.beta, [3, 2, 5]),
  };

  return response;
}

export default faker;
