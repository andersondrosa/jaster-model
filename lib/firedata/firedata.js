// import FireData from "";
import { size, isObject } from "../../src/utils";
import { jaster } from "jaster";
import { merge } from "jaster-merge";
import sha1 from "./sha1";

import MatchDir from "./directives/MatchDirective";
import OnDirective from "./directives/OnDirective";
import filters from "./directives/FiltersDirective";
import MapperDirective from "./directives/MapperDirective";

const DIRECTIVES = {
  filters: filters,
  match: MatchDir,
  on: OnDirective,
  map: MapperDirective
};

const getDirectives = (data, path, directives) => {
  //
  let clean_data = !path
    ? {
        _stages: {},
        _certifieds: {}
      }
    : {};

  for (let i in data) {
    if (i.startsWith("@")) {
      let keys = i.substring(1).split(":");

      directives.push({
        key: keys[0],
        meta: keys[1] ? keys[1] : null,
        path: path,
        data: data[i]
      });

      continue;
    }

    if (isObject(data[i])) {
      clean_data[i] = getDirectives(
        data[i],
        path + (path ? "." : "") + i,
        directives
      );

      continue;
    }

    clean_data[i] = data[i];
  }

  return JSON.parse(JSON.stringify(clean_data));
};

const getIntegrity = data => {
  return sha1.hash(JSON.stringify(data, null, 2));
};

const getSettings = (data, key, def) => {
  //
  let settings = data.hasOwnProperty("$set") ? data["$set"] : {};

  if (key === undefined) {
    return settings;
  }

  if (!settings.hasOwnProperty(key)) {
    return def == undefined ? null : def;
  }

  return settings[key];
};

const getDirectivesDefaultStage = data => {
  //
  let directives = getSettings(data, "directives");

  if (directives && directives.hasOwnProperty("default-stage")) {
    return directives["default-stage"];
  }

  return "next";
};

const getStageContents = (data, stage, module) => {
  //
  if (!data._stages) {
    return;
  }

  const sponsors = data._stages[stage];

  if (!sponsors) {
    return [];
  }

  const response = [];

  for (const i in sponsors) {
    //
    const sponsor = sponsors[i];

    if (!sponsor.hasOwnProperty(module)) {
      continue;
    }

    for (const m in sponsor[module]) {
      //
      const row = sponsor[module][m];

      response.push({
        sponsor: i,
        dir: row.dir,
        data: row.data,
        path: row.path
      });
    }
  }

  return response;
};

const match = (data, stage) => {
  //
  if (Array.isArray(stage)) {
    stage.map(stage => match(data, stage));
    return true;
  }

  if (!data._stages) {
    return;
  }

  const sponsors = data._stages[stage];

  if (!sponsors) {
    return true;
  }

  for (var i in sponsors) {
    let sponsor = sponsors[i];

    if (!sponsor.match) {
      continue;
    }

    // -----------------------------------------------------------------------
    // FILTERS

    let filters = [];

    for (let i in sponsor.filters) {
      let _filters = sponsor.filters[i].data;

      for (let key in _filters) {
        if (filters.hasOwnProperty(key)) {
          continue;
        }
        filters[key] = _filters[key];
      }
    }

    // -----------------------------------------------------------------------
    // MATCH

    let $jaster = jaster({
      filters
    });

    for (let i in sponsor.match) {
      let rules = sponsor.match[i].data;
      let _data = data;

      if (sponsor.match[i].path) {
        _data = _data[sponsor.match[i].path];
      }

      let scope = {
        scope: data,
        filters: filters
      };

      // ---------------------------------------------------------------------

      let response;

      try {
        response = $jaster.validate(_data, rules);
      } catch (e) {
        throw e;
      }

      if (!response) {
        throw "Validation Fails: " + $jaster.errors();
      }
    }
  }

  return true;
};

// ===========================================================================
// ===========================================================================

const mapData = (data, stage) => {
  //
  let response = JSON.parse(JSON.stringify(data));

  // GET FILTERS -------------------------------------------------------------

  let _filters = {};

  let filters = getStageContents(data, stage, "filters");

  for (let i in filters) {
    let data = filters[i].data;

    for (let key in data) {
      if (!_filters.hasOwnProperty(key)) {
        _filters[key] = data[key];
      }
    }
  }

  // AUTO MAPPER -------------------------------------------------------------

  let mappers = getStageContents(data, stage, "map");

  let $jaster = jaster({
    filters: _filters
  });

  let schedules = [];

  let result;

  for (let i in mappers) {
    //
    result = $jaster.map(response, mappers[i].data);

    response = result.data;

    merge(response, result.output);
  }

  return response;
};

const map = data => {
  //
  let directives = [];

  let scope = data.hasOwnProperty("$set") ? data["$set"] : {};

  let clean_data = getDirectives(data, "", directives);

  directives = mapDirectives(data, directives);

  let stages = clean_data._stages
    ? JSON.parse(JSON.stringify(clean_data._stages))
    : {};

  let stage_stage = getSettings(data, "stage", "next");

  delete stages["next"];
  delete stages[stage_stage];

  directives = JSON.parse(JSON.stringify(directives));

  merge(stages, directives);

  let response = {
    // doctype: "resource/data",
    // _stage: scope.stage ? scope.stage : "next",
    // engine: "0.1"
  };

  if (stages) {
    response._stages = stages;
  }

  response._certifieds = clean_data._certifieds ? clean_data._certifieds : {};

  response._certifieds[stage_stage] = getIntegrity(data);

  // -------------------------------------------------------------------------

  let mapped_data = mapData(clean_data, stage_stage);

  // throw 'test';

  // -------------------------------------------------------------------------
  // adicionando os modulos à resposta

  for (let module in mapped_data) {
    if (module.startsWith("$") || module.startsWith("_")) {
      continue;
    }
    if (module.match(/^[a-z][a-z0-9]*(-[a-z0-9]+)*$/g)) {
      response[module] = mapped_data[module];
    }
  }

  return response;
};

const mapDirectives = (data, directives) => {
  /*
      CONVERSÂO de Diretivas para Actions
      Aqui é executado cara componente de diretiva para cada 
      uma devolver as actions que lhe cabem.
    */
  let current_stage = getSettings(data, "stage");

  if (current_stage == null) {
    let i;
    if (!data.hasOwnProperty("_certifieds")) {
      i = 1;
    } else {
      i = size(data._certifieds) + 1;
    }

    current_stage = "stage-" + i;
  }

  let modules = [];

  for (const i in directives) {
    const row = directives[i];

    if (!DIRECTIVES.hasOwnProperty(row.key)) {
      throw 'Undefined directive "' + row.key + '"';
    }

    let directive = DIRECTIVES[row.key];

    modules.push({
      directive: row.key,
      response: directive.make(row, this)
    });
  }

  let stages = {};

  /**
   * Conversão de Actions para o padrão do "_stages"
   * stage -> sponsor -> action [] -> {scope:"", value:{}}
   * stage: Estagio atual que é executado pelo submit
   * sponsor: estágio anterior, responável por ter adicionado as ações
   * action: função a ser executada na validação ou mapeamento, ex: Match, Filters, Map
   */

  for (let n in modules) {
    let directive = modules[n].directive;

    let response = modules[n].response;

    for (let i in response) {
      let row = response[i];

      let stage_stage = row.stage ? row.stage : getDirectivesDefaultStage(data);

      if (!stages.hasOwnProperty(stage_stage)) {
        stages[stage_stage] = {};
      }

      let stage = stages[stage_stage];

      if (!stage.hasOwnProperty(current_stage)) {
        stage[current_stage] = {};
      }

      let sponsor = stage[current_stage];

      if (!sponsor.hasOwnProperty(row.action)) {
        sponsor[row.action] = [];
      }

      let actions = sponsor[row.action];

      let data = {
        data: row.data
      };

      if (row.action != directive) {
        data.dir = directive;
      }

      if (row.path) {
        data.path = row.path;
      }

      actions.push(data);
    }
  }

  return stages;
};

export { getSettings, match, map };
