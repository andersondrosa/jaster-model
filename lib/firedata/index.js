import { getSettings, match, map } from "./firedata";
import { diff } from "jaster-merge";

function submit(message) {
  //
  this.commitRequired() && this.commit("Auto commit");

  const data = this.getData();

  match(data, ["ever", getSettings(data, "stage", "next")]);

  this.put(state => diff(state, map(data)));

  this.commit(message, "submit");

  return true;
}

export default {
  install: $model => {
    if ($model.meta("using-firedata")) {
      return;
    }
    $model.meta("using-firedata", true);
    $model.mapActions({
      submit
    });
  }
};
