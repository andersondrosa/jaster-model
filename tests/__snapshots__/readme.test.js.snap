// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`README - Link and Zelda 1`] = `"Link"`;

exports[`README - Link and Zelda 2`] = `"villager"`;

exports[`README - Link and Zelda 3`] = `
Object {
  "gender": "male",
  "name": "Link",
  "rupees": 130,
  "type": "warrior 🧝🏼‍♂",
}
`;

exports[`README - Link and Zelda 4`] = `
Array [
  Object {
    "data": Object {
      "name": "Link",
      "rupees": 130,
      "type": "warrior 🧝🏼‍♂",
    },
    "integrity": "de9148ce7c6ee82fa253b3975364c33bd669f9b7",
    "type": "put",
  },
]
`;

exports[`README - Link and Zelda 5`] = `"Evil 'Link' 🧟‍♂"`;

exports[`README - Link and Zelda 6`] = `"Link"`;

exports[`README - Link and Zelda 7`] = `
Object {
  "gender": "male",
  "name": "Character",
  "rupees": 30,
  "type": "villager",
}
`;

exports[`README - Link and Zelda 8`] = `
Array [
  Object {
    "data": Object {
      "name": "Link",
      "rupees": 130,
      "type": "warrior 🧝🏼‍♂",
    },
    "integrity": "de9148ce7c6ee82fa253b3975364c33bd669f9b7",
    "type": "put",
  },
]
`;

exports[`README - Link and Zelda 9`] = `
Object {
  "gender": "male",
  "name": "Link",
  "rupees": 130,
  "type": "warrior 🧝🏼‍♂",
}
`;

exports[`README - Link and Zelda 10`] = `
Object {
  "gender": "female",
  "name": "Zelda",
  "rupees": 2600,
  "type": "princess 🧝🏼‍♀",
  "wholoves": Object {
    "name": "Link ♥",
    "type": "warrior 🧝🏼‍♂ 💪🏻",
  },
}
`;

exports[`README - Link and Zelda 11`] = `
Object {
  "commits": Array [
    Object {
      "data": Object {
        "name": "Link",
        "rupees": 130,
        "type": "warrior 🧝🏼‍♂",
      },
      "integrity": "de9148ce7c6ee82fa253b3975364c33bd669f9b7",
      "type": "put",
    },
  ],
  "data": Object {
    "gender": "male",
    "name": "Character",
    "rupees": 30,
    "type": "villager",
  },
}
`;

exports[`README - Link and Zelda 12`] = `
Object {
  "commits": Array [
    Object {
      "data": Object {
        "name": "Link",
        "rupees": 130,
        "type": "warrior 🧝🏼‍♂",
      },
      "integrity": "de9148ce7c6ee82fa253b3975364c33bd669f9b7",
      "type": "put",
    },
    Object {
      "data": Object {
        "gender": "female",
        "name": "Zelda",
        "rupees": 2600,
        "type": "princess 🧝🏼‍♀",
        "wholoves": Object {
          "name": "Link ♥",
          "type": "warrior 🧝🏼‍♂ 💪🏻",
        },
      },
      "integrity": "19f81578efb0e1f7ecda8b2f2ff01b67e5829c93",
      "type": "put",
    },
  ],
  "data": Object {
    "gender": "male",
    "name": "Character",
    "rupees": 30,
    "type": "villager",
  },
}
`;
